File: figs/oplog.lua
package.path = package.path .. ';lib/?.lua'
local luamp = require 'luamp'
local stream = require 'stream'
local table = require 'table'
local string = require 'string'

local SEG_WIDTH = 1
local SEG_HEIGHT = 0.8

local rects = stream
    .iterate(
        luamp.origin,
        function(x)
            return x + luamp.point(SEG_WIDTH, 0)
        end)
    :take(5)
    :map(function(x)
        return luamp.rectangle(x, SEG_WIDTH, SEG_HEIGHT)
    end)
    :collect()

local figs = {}

stream.from_list(rects)
    :drop(2)
    :map(function(x)
        local vertices = x:vertices()
        return luamp.line(vertices[1], vertices[4])
    end)
    :collect(figs)

stream.zip(
    stream.from_list({'n-2', 'n-1', 'n'}),
    stream.from_list(rects):drop(2))
    :map(function(x)
        local idx = x[1]
        local rect = x[2]
        return stream.from_list({
            luamp.text(
                rect:center(),
                luamp.directions.center,
                string.format('$\\text{op}_{%s}$', idx)),
            luamp.text(
                rect:vertices()[1],
                luamp.directions.top,
                string.format('$\\text{SN}_{%s}$', idx))})
    end)
    :flatten()
    :collect(figs)

local vs = {
    rects[1]:vertices()[1],
    rects[5]:vertices()[2],
    rects[5]:vertices()[3],
    rects[1]:vertices()[4]}
stream
    .zip(
        stream.from_list(vs),
        stream.from_list(vs)
            :chain(stream.from_list(vs))
            :drop(1))
    :map(function(x)
        return luamp.line(x[1], x[2])
    end)
    :collect(figs)

local t = luamp.text(
    luamp.centroid(rects[1]:vertices()[2], rects[1]:vertices()[3]),
    luamp.directions.center,
    '\\dots')
table.insert(figs, t)

print(luamp.figure(table.unpack(figs)))

File: figs/oplog.py
from fathom import Point, ORIGIN
from fathom.tikz import Canvas
import fathom.geometry as geo
import fathom.tikz.colors as colors
import fathom.tikz.line_styles as line_styles
import fathom.tikz.locations as locations
from itertools import *

SEG_WIDTH = 1
SEG_HEIGHT = 0.8

if __name__ == '__main__':
    rects = repeat(ORIGIN, 5)
    rects = accumulate(rects, lambda x, _: x + geo.Point(SEG_WIDTH, 0))
    rects = [geo.Rectangle(center=x, width=SEG_WIDTH, height=SEG_HEIGHT) for x in rects]

    canvas = Canvas(
        preamble=['\\usepackage{amsmath}'],
        leading_instructions=['\\footnotesize'])

    ops = ['$\\text{op}_{n-2}$', '$\\text{op}_{n-1}$', '$\\text{op}_{n}$']
    sns = ['$\\text{SN}_{n-2}$', '$\\text{SN}_{n-1}$', '$\\text{SN}_{n}$']
    for x, op, sn in zip(rects[2:], ops, sns):
        canvas.new_text(anchor=x.center(), text=op)
        canvas.new_text(
            anchor=x.vertices()[0],
            text=sn,
            location=locations.NORTH)

    canvas.new_text(
        anchor=geo.centroid([rects[0].center(), rects[1].center()]),
        text='\\normalsize \\dots')

    canvas.new_rectangle(
        vertices=[
            rects[0].vertices()[0],
            rects[-1].vertices()[1],
            rects[-1].vertices()[2],
            rects[0].vertices()[3]])
    for x in rects[2:]:
        vs = x.vertices()
        canvas.new_line(src=vs[0], dst=vs[3])

    print(canvas.draw())
